### PR TITLE
Implemented db_map

### LINKED ISSUE #2: Database for communication
Create basic class of the database for simpler communication.

Use this library to establish connection with [**firebase database**](https://github.com/ozgur/python-firebase).

Implement this functions:

**AUTHENTIFICATION**
- [x] add_new_user()
- [x] check_user_password()

**MAP PAGE**
- [x] get_cities()
- [x] get_categories()
- [x] get_receiver_station_types()
- [x] get_receive_stations()

### PR DESCRIPTION
**Pull request by yarmor20**
Closes #2

### CODE REVIEW

File: database/db_map.py [hunk 1]
```diff
@@ -1,29 +1,28 @@
 class DBMap:
     def __init__(self, database):
         self.db = database
-
-        # TODO: Add link to some tables
+        # super(DBMap, self).__init__()
```
Comment (sofiia-tesliuk) [inline database/db_map.py]: Please, change to:
super().__init__(database)

File: database/db_map.py [hunk 2]
```diff
@@ -1,29 +1,28 @@
...
 
     def get_cities(self):
```
Comment (sofiia-tesliuk) [inline database/db_map.py:8]: :thumbsup:

File: database/db_map.py [hunk 3]
```diff
@@ -1,29 +1,28 @@
...
         """
         Return all available cities in database.
         :return: (dict) {city_id: "city_name", ...}
         """
-        raise NotImplementedError
+        return self.db.get('cities', None)
 
     def get_categories(self):
```
Comment (sofiia-tesliuk) [inline database/db_map.py:15]: :thumbsup:

File: database/db_map.py [hunk 4]
```diff
@@ -1,29 +1,28 @@
...
         """
         Return all available categories in database.
         :return: (dict) {category_id: "category_name", ...}
         """
-        raise NotImplementedError
+        return self.db.get('categories', None)
 
     def get_receiver_station_types(self):
```
Comment (sofiia-tesliuk) [inline database/db_map.py:22]: :thumbsup:

File: database/db_map.py [hunk 5]
```diff
@@ -43,4 +41,74 @@ def get_receive_stations(self, rcv_station_dict):
...
+        suitable_stations = []
+        for station in stations:
+            if self._data_intersection(user_station['categories'], station['categories']):
+                if self._data_intersection(user_station['city_id'], station['locations']):
+                    if self._data_intersection(user_station['receive_station_type'], station['type_id']):
+
+                        s_dates = self._get_all_dates(station['time_from'], station['time_to'])
+                        u_dates = self._get_all_dates(user_station['time_from'], user_station['time_to'])
+                        if self._data_intersection(s_dates, u_dates):
+                            suitable_stations.append(station)
```
Comment (sofiia-tesliuk) [inline database/db_map.py]: Way too complex option to see if there is dates intersection. Specially when you call function self._get_all_dates(user_station['time_from'], user_station['time_to']) each time and receive the same output.

You can check if there `not ((user_station['time_from'] > station['time_to']) or (user_station['time_to'] < station['time_from']))`

File: database/db_map.py [hunk 6]
```diff
@@ -43,4 +41,74 @@ def get_receive_stations(self, rcv_station_dict):
...
+        end_date_s = date(s_time_to[0], s_time_to[1], s_time_to[2])  # end date
+
+        delta = end_date_s - start_date_s  # as timedelta
+
+        available_dates = []
+        for i in range(delta.days + 1):
+            day = start_date_s + timedelta(days=i)
+            available_dates.append(day)
+
+        return available_dates
```
Comment (sofiia-tesliuk) [inline database/db_map.py]: Better to have date comparison:
```
def compare_dates(date1, date2):
    if date1 < date2 return 1
    if date1 == date2 return 0
    if date1 > date2 return -1
```

File: database/db_map.py [hunk 7]
```diff
@@ -43,4 +41,74 @@ def get_receive_stations(self, rcv_station_dict):
                                                     categories: list(int),
                                                     items: list(int)}, ...}
         """
-        raise NotImplementedError
+        response = self.db.get("receiver_stations", None)
+        self.checking(rcv_station_dict, response)
+        return response
+
+    def checking(self, user_station, stations):
+        """
```
Comment (sofiia-tesliuk) [inline database/db_map.py:50]: Better to check inside the function only one element if it suits user_info. Iterate over stations in `get_receive_stations(...)`

File: database/db_map.py [hunk 8]
```diff
@@ -43,4 +41,74 @@ def get_receive_stations(self, rcv_station_dict):
...
+        Return all suitable for a user stations,
+        otherwise return []
+        :param user_station: dict = {(str): int, (str): (str), ...}
+        :param stations:  dict = {(str): list, (str): (int), (str): (str), ...}
+        :return: list = [(dict), ...]
+        """
+        suitable_stations = []
+        for station in stations:
+            if self._data_intersection(user_station['categories'], station['categories']):
+                if self._data_intersection(user_station['city_id'], station['locations']):
```
Comment (sofiia-tesliuk) [inline database/db_map.py]: You are comparing here city_id and location_id. Which is not really what is needed.
Create function `get_location(location_id)` from which you can extract location info with city_id.

File: database/db_map.py [hunk 9]
```diff
@@ -43,4 +41,74 @@ def get_receive_stations(self, rcv_station_dict):
...
+                    if self._data_intersection(user_station['receive_station_type'], station['type_id']):
```
Comment (sofiia-tesliuk) [inline database/db_map.py]: `station['type_id']` -- is just one element.
Just check `if station['type_id'] in user_station['receive_station_type']`.

File: database/db_map.py [hunk 10]
```diff
@@ -43,4 +41,74 @@ def get_receive_stations(self, rcv_station_dict):
...
+        if not isinstance(database_data, list):
+            database_data = [database_data]
+
+        user_set = set(user_data)
+        database_set = set(database_data)
+
+        if user_set.intersection(database_set) != set():
+            return True
+        else:
+            return False
```
Comment (sofiia-tesliuk) [inline database/db_map.py]: Set intersection need to be used only for categories. Maybe there is no need in one more additional function.

**Review by sofiia-tesliuk — APPROVED** [review]

**Review by LAnastasiia — APPROVED** [review]
Fantastic :)

**Pull request merged**